Mastodon blog post (#7)
julianocosta89/jcosta.dev: content/posts/mastodon-otel/index.md

### The Collector

As Mastodon wanted to allow users to send telemetry data to any observability backend, I wanted to build a minimal collector,
containing just the components required for each vendor, and each developer would configure their Collector in their own way.

To do that, I've used a multi-staged Dockerfile which first would use the OCB (OpenTelemetry Collector Builder) to build a
minimal Collector distro based on a manifest file that specifies which components should be part of the Collector distro. And
then runs the Collector with its configuration file.

Here is how a `manifest.yaml` file looks like:

{{% snippet manifest.yaml yaml %}}

This is how the `Dockerfile` would look like:

{{% snippet Dockerfile Dockerfile %}}

And finally, here is how the `otelcol-config.yaml` file would look like:

{{% snippet otelcol-config.yaml yaml %}}

{{< alert >}}
Note that you can only use the components previously added to the manifest file, as the custom Collector has only
those components. If you try to use a component that is not defined in the manifest the Collector will fail to start.
{{< /alert >}}

> In the PR, I also added a `TELEMETRY_BACKEND` build arg, which would get different manifest files depending on

Author: ```suggestion
In the PR, I also added a `TELEMETRY_BACKEND` build arg, which would use a different manifest file depending on
```

which observability backend the user wanted to send data to. The Datadog manifest had the Datadog components,
but the default had only the `OTLP` and vendor agnostic components.

It was built to allow users to collaborate and add their manifests with their required components later on.

If you want to know more about the OCB, you can take a look at the official OpenTelemetry docs: [Building a custom collector][].

### The Instrumentation

The main difference between the initial PR from Robb and mine in the instrumentation side, was that instead of using the
`opentelemetry-instrumentation-all`, I suggested adding only the dependencies we need.

This is way more verbose and requires more maintenance, but we have more control over what we are instrumenting. This also
aligned with the idea behind building a minimal Collector distro.

We don't import everything automatically, we define what we want, what we need, and only use those.

## The Collector Configuration

> After the PR from the Mastodon team got merged, [Renaud Chaput] and [Tim Campbell] started

Author: ```suggestion
After the PR from the Mastodon team got merged, [Renaud Chaput][] and [Tim Campbell][] started
```

points and troubles when sending OTel data to Datadog is one of the most important part of my job.

In the [Datadog's Open Source Hub] blog post I've spent some time talking about [Inferred Services]
and how that improved Mastodon's usability significantly.

In here, I want to focus on some Collector configuration that helped them navigate better the
OTel world in Datadog.

### Resource Processor

In the snippet below, the [resourceprocessor] component is used to set an attribute
`deployment.environment.name=development`.

When the OTel data arrives at Datadog with that attribute, the service is tagged with the
defined value, and that makes their life easier when navigating the data in Datadog.

```yaml
processors:
  [...]
  resource:
    attributes:
      - key: deployment.environment.name
        value: "development"
        action: upsert
```

### Transform Processor

Working with the [transformprocessor] was a great learning to me. I knew the existence of the component,
but I have never used it myself.

Mastodon wanted some help to get a better naming to their spans, other than the default Ruby instrumentation name.

> They would like to have the values from `code.namespace` be merged with `code.function`. If you don't know

Author: ```suggestion
They would like to have the values from `code.namespace` merged with `code.function`. If you don't know
```

Another talk that I have to mention is from [Vijay and Aishwarya from eBay] in this year's (2024) Observability
Day EU, where they blew my mind when said that they sample just 1% of their traces, as that percentage is
enough for their traffic volume 🤯.

Tail sampling has a catch when you use multiple Collectors, where you have to make sure all spans from
a trace reaches the same Collector, otherwise you will have some broken traces.

In Mastodon's case though, I got lucky because they are using a single instance of the Collector, which
made our life configuring it way easier.

Initially I've suggested 25% of sampling, plus all errors, but Tim was already satisfied with 10% plus errors.

Below we have the `tail_sampling` processor definition for this.

```yaml
processors:
  [...]
  tail_sampling:
    policies:
      [
        {
          name: errors-policy,
          type: status_code,
          status_code: {status_codes: [ERROR]}
        },
        {
          name: randomized-policy,
          type: probabilistic,
          probabilistic: {sampling_percentage: 10}
        },
      ]
```

Another point to highlight about tail sampling is that if you calculate metrics from spans

> with components such as the [spanmetricsconnector] or [datadogconnector], you have to make sure

Author: ```suggestion
with components such as the [spanmetricsconnector][] or [datadogconnector][], you have to make sure
```

the metrics are being calculated before the sampling.

Let me elaborate a bit on that.

> Imagine your service emits 100 traces and you have a tail sampling configure to 25%.

Author: ```suggestion
Imagine your service emits 100 traces and you have a tail sampling configured to 25%.
```
